### ISSUE TITLE
support EIP-1271 in permit

### ISSUE THREAD

**Issue by moodysalem** [WETH10/WETH10#62]
https://github.com/ethereum/EIPs/issues/1271

This would allow contract signatures to be used for WETH10

**Comment by albertocuestacanada** [comment]
Are there any live implementations of this?

**Comment by moodysalem** [comment]
> Are there any live implementations of this?

yes, argent here https://github.com/argentlabs/argent-contracts/blob/cd2deb75ba76721ec797eae2dae8b383e4578a1f/contracts/modules/TransferManager.sol

iirc also gnosis safe implements it

or do you mean examples of supporting it in the token?

**Comment by albertocuestacanada** [comment]
My main concern is that despite it being a good idea, and there being a few implementations of it (Gnosis, Argent, 0x), I can't find any examples of using it inside `permit`.

If you are familiar with how to implement EIP1271 and would like to draft a PR, maybe we can add it to WETH10.

**Comment by moodysalem** [comment]
> My main concern is that despite it being a good idea, and there being a few implementations of it (Gnosis, Argent, 0x), I can't find any examples of using it inside `permit`.
> 
> 
> 
> If you are familiar with how to implement EIP1271 and would like to draft a PR, maybe we can add it to WETH10. 

should be as simple as this for the token (using the oz address library)


using Address for address;
...

if (owner.isContract()) {
 // call eip 1271 interface
} else {
 // normal signature verification
}

**Comment by albertocuestacanada** [comment]
This still falls short of something I can implement with confidence.

We would need, at the very least:
1. An example of how to create the signatures, for testing.
2. Conversion of the [EIP1271 logic](https://github.com/argentlabs/argent-contracts/blob/cd2deb75ba76721ec797eae2dae8b383e4578a1f/contracts/modules/TransferManager.sol#L487) into the ERC2612 format, using deadline and nonces.

Someone familiar with the standard should step in

**Comment by moodysalem** [comment]
> * An example of how to create the signatures, for testing.

The signatures can be whatever you want for testing, e.g. a constant. The verifying contract just checks the signature and effectively returns a boolean (the specific bytes4 value from the spec)

> Conversion of the EIP1271 logic into the ERC2612 format, using deadline and nonces.

IIUC ERC2612 just relates to producing the msg hash that is signed, while the contract handles validating the signature.

**Comment by alfredolopez80** [comment]
The best approach for that Based on Open Zeppelin Library is SignatureChecker in the version 4.1 
https://docs.openzeppelin.com/contracts/4.x/api/utils#SignatureChecker

**Comment by albertocuestacanada** [comment]
I don't get it, how do you implement this EIP?

Something like the code below?

```
    function permitEip1271(
        address signer,
        address owner,
        address spender,
        uint256 value,
        bytes32 hash,
        bytes memory signature
    ) external override {

       require(
                IERC1271(signer).isValidSignature(hash, signature)  == IERC1271(signer).isValidSignature.selector,
                "Invalid signature"
        );

        // _approve(owner, spender, value);
        allowance[owner][spender] = value;
        emit Approval(owner, spender, value);
    }
```

If so, how does the `value` get verified?
No deadline either?
Are the `signer` and the `owner` the same account?
What is the name of this function?
If this is to be part of `permit`, how do we reconcile that the parameters are different?

Also, we already messed up by including ERC667 in draft form in WETH10, I'm not too keen on including EIP1271 in WETH8 while in draft form.
